### ISSUE TITLE
PFPKI: Cloud integration with Intune not working

### ISSUE THREAD

**Issue by nathanbloemendaal78** [inverse-inc/packetfence#6933]
**Describe the bug**
Using the cloud integration option in  SCEP settings in a PKI template seems to break SCEP functionality for that template.

**To Reproduce**
Configure the PKI as per the steps in the manual. Configure an Intune cloud connector. Do not enable this connector in the SCEP settings in the template:

![image](https://user-images.githubusercontent.com/62255038/158386652-25991ea7-b54e-4b7b-a161-faef88134026.png)

SCEP functionality works, both locally and on a Windows 10 laptop:

![image](https://user-images.githubusercontent.com/62255038/158401094-1410ce23-8fd6-4121-8f03-07a7e678c935.png)

Turn on the cloud integration in the template:

![image](https://user-images.githubusercontent.com/62255038/158386933-96c09366-cac1-459e-99d5-3bff05fec8c5.png)

SCEP functionality is broken, both locally and on a Windows 10 laptop:

![image](https://user-images.githubusercontent.com/62255038/158400875-d8986e2e-397b-4e3f-9620-ec48d622d508.png)

**Expected behavior**
Cloud functionality should work on a template, to be able to provision Intune managed devices with certificates to be used with EAP-TLS.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser Chrome
 - Version 99

**Comment by fdurand** [comment]
Do you have the logs on the PacketFence side when you do the scep request ?

**Comment by nathanbloemendaal78** [comment]
Hi,

yes, this is how far a SCEP request gets when initiated from a Windows 10 laptop:

![image](https://user-images.githubusercontent.com/62255038/158579565-1b9945d7-6cd2-4c14-8f15-a37e768a6573.png)

If I initiate a sscep get ca on the packetfence instance, result is the same. If I initiate a scep get ca with cloud integration turned off, I already get a bit further, same result for the Windows 10 laptop:

![image](https://user-images.githubusercontent.com/62255038/158580115-2a523e39-d56e-49a7-9800-d3dae370f49a.png)

I am not at my test environment right now, I can do more tests or gather more logs tomorrow.

**Comment by nathanbloemendaal78** [comment]
@fdurand are more logs required?

**Comment by fdurand** [comment]
yes if it´s possible to do a journalctl -f when you do the request in both cases.
Thanks

**Comment by nathanbloemendaal78** [comment]
Cloud connector disabled on template:

Mar 21 16:00:08 dev-eduguard02 pfpki[53972]: t=2022-03-21T16:00:08+0100 lvl=info msg="Got GET request from 127.0.0.1:32898" pid=53972
Mar 21 16:00:08 <hostname> pfpki[53972]: t=2022-03-21T16:00:08+0100 lvl=info msg="SCEP GET To: /api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACaps&message=default" pid=53972
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.740338969Z caller=service_logging.go:22 component=scep_service method=GetCACaps err=null took=637ns
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.740892706Z caller=endpoint.go:186 op=GetCACaps error=null took=554.615Âµs
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.740912349Z caller=logutil.go:70 component=http method=GET status=200 proto=HTTP/1.1 host=127.0.0.1 user_agent="Mozilla/4.0 (compatible; Win32; NDES client 10.0.19041.1466/vb_release_svc_prod1)" path="/api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACaps&message=default"
Mar 21 16:00:08 <hostname> haproxy[5143]: 157.97.212.186:57525 [21/Mar/2022:16:00:08.737] portal-http-10.7.1.122 pki/127.0.0.1 0/0/0/3/3 200 181 - - ---- 2/1/0/0/0 0/0 {157.97.212.165} "GET /scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACaps&message=default HTTP/1.1"
Mar 21 16:00:08 <hostname> pfpki[53972]: t=2022-03-21T16:00:08+0100 lvl=info msg="Got GET request from 127.0.0.1:32900" pid=53972
Mar 21 16:00:08 <hostname> pfpki[53972]: t=2022-03-21T16:00:08+0100 lvl=info msg="SCEP GET To: /api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACert&message=default" pid=53972
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.76427514Z caller=service_logging.go:34 component=scep_service method=GetCACert message=default err=null took=488ns
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.764313033Z caller=endpoint.go:186 op=GetCACert error=null took=39.805Âµs
Mar 21 16:00:08 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:08.764330633Z caller=logutil.go:70 component=http method=GET status=200 proto=HTTP/1.1 host=127.0.0.1 user_agent="Mozilla/4.0 (compatible; Win32; NDES client 10.0.19041.1466/vb_release_svc_prod1)" path="/api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACert&message=default"
Mar 21 16:00:08 <hostname> haproxy[5143]: 157.97.212.186:57525 [21/Mar/2022:16:00:08.761] portal-http-10.7.1.122 pki/127.0.0.1 0/0/0/2/2 200 1119 - - ---- 2/1/0/0/0 0/0 {157.97.212.165} "GET /scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACert&message=default HTTP/1.1"
Mar 21 16:00:09 <hostname> pfpki[53972]: t=2022-03-21T16:00:09+0100 lvl=info msg="Got POST request from 127.0.0.1:32904" pid=53972
Mar 21 16:00:09 <hostname> pfpki[53972]: t=2022-03-21T16:00:09+0100 lvl=info msg="SCEP POST To: /api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=PKIOperation" pid=53972
Mar 21 16:00:09 <hostname> pfhttpd[53972]: ts=2022-03-21T15:00:09.079558437Z caller=service.go:89 msg="failed to sign CSR" err="invalid challenge"
Mar 21 16:00:09 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:09.085560698Z caller=service_logging.go:47 component=scep_service method=PKIOperation err=null took=9.964569ms
Mar 21 16:00:09 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:09.085614403Z caller=endpoint.go:186 op=PKIOperation error=null took=10.020146ms
Mar 21 16:00:09 <hostname> pfhttpd[53972]: level=info ts=2022-03-21T15:00:09.085649688Z caller=logutil.go:70 component=http method=POST status=200 proto=HTTP/1.1 host=127.0.0.1 user_agent="Mozilla/4.0 (compatible; Win32; NDES client 10.0.19041.1466/vb_release_svc_prod1)" path="/api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=PKIOperation"
Mar 21 16:00:09 <hostname> haproxy[5143]: 157.97.212.186:57525 [21/Mar/2022:16:00:09.071] portal-http-10.7.1.122 pki/127.0.0.1 0/0/0/14/14 200 1847 - - ---- 2/1/0/0/0 0/0 {157.97.212.165} "POST /scep/edgdevicecertificate7/pkiclient.exe?operation=PKIOperation HTTP/1.1"

cloud connector enabled on template:

Mar 21 16:16:00 <hostname> pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.463083646Z caller=scep.go:63 err="Enable to create Cloud service"
Mar 21 16:16:00 <hostname> pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.463117378Z caller=scep.go:64 err="Insufficient privileges to complete the operation."
Mar 21 16:16:00 <hostname> haproxy[5143]: 157.97.212.186:57563 [21/Mar/2022:16:16:00.201] portal-http-10.7.1.122 pki/127.0.0.1 0/0/0/261/261 200 98 - - ---- 2/1/0/0/0 0/0 {157.97.212.165} "GET /scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACaps&message=default HTTP/1.1"
Mar 21 16:16:00 <hostname> pfpki[903819]: t=2022-03-21T16:16:00+0100 lvl=info msg="Got GET request from 127.0.0.1:37702" pid=903819
Mar 21 16:16:00 <hostname> pfpki[903819]: t=2022-03-21T16:16:00+0100 lvl=info msg="SCEP GET To: /api/v1/scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACert&message=default" pid=903819
Mar 21 16:16:00 <hostname> pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.761846867Z caller=scep.go:63 err="Enable to create Cloud service"
Mar 21 16:16:00 <hostname> pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.761887833Z caller=scep.go:64 err="Insufficient privileges to complete the operation."
Mar 21 16:16:00 <hostname> haproxy[5143]: 157.97.212.186:57563 [21/Mar/2022:16:16:00.481] portal-http-10.7.1.122 pki/127.0.0.1 0/0/0/280/280 200 98 - - ---- 2/1/0/0/0 0/0 {157.97.212.165} "GET /scep/edgdevicecertificate7/pkiclient.exe?operation=GetCACert&message=default HTTP/1.1"

**Comment by nathanbloemendaal78** [comment]
@fdurand see above; I found one typo in my config, these logs are from after the correction. Any ideas?

**Comment by fdurand** [comment]
@nathanbloemendaal78, when the cloud connector is enabled , it´s a permission issue in azure:

Mar 21 16:16:00 pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.761846867Z caller=scep.go:63 err="Enable to create Cloud service"
Mar 21 16:16:00 pfhttpd[903819]: level=info ts=2022-03-21T15:16:00.761887833Z caller=scep.go:64 err="Insufficient privileges to complete the operation."

If the cloud is disabled then it´s normal that the challenge fail since pfpki idoesn´t know what is the challenge sent by intune.

https://docs.microsoft.com/en-us/mem/intune/protect/scep-libraries-apis

So for me your issue is on the azure part.
Check to see if the permissions matches: https://www.packetfence.org/doc/PacketFence_Installation_Guide.html#_azure_configuration

If it´s the case then they probably changed something in the azure side (it´s already happened).

**Comment by nathanbloemendaal78** [comment]
My apologies for not getting back to you sooner. In the meantime, I set up a fresh PF 11.2 install /w patches and a new Azure tenant to test with. I also switched to using a Win10 VM to test with. Unfortunately, I ended up with exactly the same errors... I checked Azure audit logs for the Service Principal used, and all calls seem to be succesfull. Could it be that something goes wrong in an API call to either Azure or Intune in steps 4,5 or 6 below:

![image](https://user-images.githubusercontent.com/62255038/162954458-0e006f39-9b51-41bd-b20d-d4ae4283dc82.png)

I would very much like to increase the debug level in PacketFence for the processes involved, but I can't seem to get that to work. 
Any new ideas and/or suggestions welcome!

**Comment by trs80** [comment]
It was pleasingly easy to edit intune.go and add logging. I updated it to use the Graph API instead (since assigning Azure AD API permissions isn't easy):
https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/end-of-support-for-azure-ad-graph-permission-sign-up-through/ba-p/2464404
https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/azure-ad-change-management-simplified/ba-p/2967456
https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-request-differences

by changing these:
const graphApiVersion = "v1.0"
const graphResourceUrl = "https://graph.microsoft.com/"

and in NewCloud():
        graphRequest := graphResourceUrl + graphApiVersion + "/servicePrincipal/appId=" + intuneAppId + "/endpoints"

(I'll make a proper diff later)

That got me to this error which I will look into tomorrow:

Aug 16 16:31:28 kerr pfhttpd[260172]: ts=2022-08-16T08:31:28.075441935Z caller=service.go:89 msg="failed to sign CSR" err="Post \"/ScepActions/validateRequest\": unsupported protocol scheme \"\""

**Comment by nathanbloemendaal78** [comment]
Good to see somebody investigates! I had parked this for a while, but I will make the changes on my end as well.

**Comment by trs80** [comment]
You may have initially been hitting https://github.com/inverse-inc/packetfence/pull/7021 as well, which was only fixed in June (and pushed to the 11.2 branch).

One more change is needed when moving to MS Graph API, searching for providerName instead of serviceName and some attribute changes. Diff attached, it'll need some more work before merging since it should handle old (AAD Graph) and new (MS Graph) instead of just new. But it was enough to issue a certificate to my test device.

https://github.com/microsoft/Intune-Resource-Access/blob/develop/src/CsrValidation/java/lib/src/main/java/com/microsoft/intune/scepvalidation/IntuneClient.java#L387

**Comment by trs80** [comment]
Can't attach a diff? Weird, here it is:
```diff -Nru ./caddy/pfpki/cloud/intune.go /usr/local/pf/go/caddy/pfpki/cloud/intune.go
--- ./caddy/pfpki/cloud/intune.go   2022-08-17 10:39:54.581057260 +0800
+++ /usr/local/pf/go/caddy/pfpki/cloud/intune.go    2022-08-17 10:37:13.910029723 +0800
@@ -81,8 +83,10 @@

 const intuneAppId = "0000000a-0000-0000-c000-000000000000"
 const intuneResourceUrl = "https://api.manage.microsoft.com/"
-const graphApiVersion = "1.6"
-const graphResourceUrl = "https://graph.windows.net/"
+//const graphApiVersion = "1.6"
+//const graphResourceUrl = "https://graph.windows.net/"
+const graphApiVersion = "v1.0"
+const graphResourceUrl = "https://graph.microsoft.com/"

 func NewIntuneCloud(ctx context.Context, name string) (Cloud, error) {

@@ -134,8 +157,8 @@
        token = spt.Token()
        Bearer = "Bearer " + token.AccessToken
    }

-   graphRequest := graphResourceUrl + cl.TenantID + "/servicePrincipalsByAppId/" + intuneAppId + "/serviceEndpoints?api-version=" + graphApiVersion
+   graphRequest := graphResourceUrl + graphApiVersion + "/servicePrincipals/appId=" + intuneAppId + "/endpoints"

    tr := &http.Transport{
        TLSClientConfig: &tls.Config{CipherSuites: []uint16{
@@ -195,15 +227,13 @@
        if k == "value" {
            for _, n := range v.([]interface{}) {
                for a, b := range n.(map[string]interface{}) {
-                   if a == "serviceName" {
+                   if a == "providerName" {
                        if b == VALIDATION_SERVICE_NAME {
-                           apiEndpoint.OdataType = n.(map[string]interface{})["odata.type"].(string)
-                           apiEndpoint.ObjectId = n.(map[string]interface{})["objectId"].(string)
-                           apiEndpoint.ResourceId = n.(map[string]interface{})["resourceId"].(string)
-                           apiEndpoint.ObjectType = n.(map[string]interface{})["objectType"].(string)
+                           apiEndpoint.ObjectId = n.(map[string]interface{})["id"].(string)
+                           apiEndpoint.ResourceId = n.(map[string]interface{})["providerResourceId"].(string)
                            apiEndpoint.Capability = n.(map[string]interface{})["capability"].(string)
-                           apiEndpoint.ServiceId = n.(map[string]interface{})["serviceId"].(string)
-                           apiEndpoint.ServiceName = n.(map[string]interface{})["serviceName"].(string)
+                           apiEndpoint.ServiceId = n.(map[string]interface{})["providerId"].(string)
+                           apiEndpoint.ServiceName = n.(map[string]interface{})["providerName"].(string)
                            apiEndpoint.Uri = n.(map[string]interface{})["uri"].(string)
                        }
                    }
```

**Comment by trs80** [comment]
You should probably also look at switching to MSAL instead of ADAL for the MS Graph calls since ADAL is going away by the end of the year https://techcommunity.microsoft.com/t5/azure-active-directory-identity/update-your-applications-to-use-microsoft-authentication-library/ba-p/1257363

**Comment by trs80** [comment]
Once this lands you can email Lance Crandall ( [lacranda@microsoft.com](mailto:lacranda@microsoft.com) ) to get put on the list of third-party CAs that support Intune SCEP per https://github.com/MicrosoftDocs/memdocs/pull/3185
